### PR TITLE
add operation message to cordon/uncordon/drain

### LINKED ISSUE #55527: Add comments info when kubectl drain/cordon/uncordon
**Is this a BUG REPORT or FEATURE REQUEST?**:
> /kind feature

**What happened**:
For now, when we `cordon`/`uncordon`/`drain` a node from a k8s cluster, we can not add a comment describing the reason to do so. An comment is useful for recording the reason in case someone else want to know the root cause.
**What you expected to happen**:
when running `kubectl cordon/uncordon/drain NODE`, we can add a comment and running like this: `kubectl cordon/uncordon/drain NODE COMMENT`

/cc @kubernetes/sig-node-feature-requests

### PR DESCRIPTION
**Pull request by andyxning**
Fix #55527 

This PR will add a `--message` command line argument to kubectl `cordon/uncordon/drain` to add a simple description message about the `cordon/uncordon/drain` operation.

And, this will add a `MESSAGE` column to `kubectl get nodes` ouput.

Usage:
```
kubectl cordon 192.168.0.1 -m "disk is full"
node "192.168.0.1" cordoned

kubect get nodes
NAME          STATUS     ROLES     AGE       VERSION                   MESSAGE
192.168.0.1  Ready       <none>    1d           v1.8.0                        disk is full

kubect get node 192.168.0.2
NAME          STATUS     ROLES     AGE       VERSION                   MESSAGE
192.168.0.2  Ready       <none>    1d           v1.8.0                        <none>
```

**Release note**:
```release-note
Add description message to corond/uncordon/drain
```

### CODE REVIEW

**Comment by weekface** [comment]
Add this message to node's `Annotations` may be better.

**Comment by andyxning** [comment]
@weekface 
> Add this message to node's Annotations may be better.

I think we also need the first and quick display about the message when `cordon/uncordon/drain` a node. With `Annotations` we need to make a constraint about the key to associate the message.  We can not gurantee that the specified key does not already used by others.

File: pkg/kubectl/cmd/drain.go [hunk 1]
```diff
@@ -113,6 +114,7 @@ func NewCmdCordon(f cmdutil.Factory, out io.Writer) *cobra.Command {
 		},
 	}
 	cmd.Flags().StringVarP(&options.Selector, "selector", "l", options.Selector, "Selector (label query) to filter on")
+	cmd.Flags().StringVarP(&options.Message, "message", "m", options.Message, "Message describing the reason to cordon a node")
```
Comment (shiywang) [inline pkg/kubectl/cmd/drain.go]: according to [cli convention](https://github.com/kubernetes/community/blob/master/contributors/devel/kubectl-conventions.md#flag-conventions):

> Flag names and single-character aliases should have the same meaning across all commands

Comment (andyxning) [reply pkg/kubectl/cmd/drain.go]: @shiywang 

I am sorry for not understanding your suggestions. Could you please detail it.
FYI: `-m` or `--message` is added according to `git commit -m` and `docker commit -m`.

Comment (shiywang) [reply pkg/kubectl/cmd/drain.go]: I mean maybe we should first not having shortcut of this flag because once we add it, it should have same meaning across all kubectl commands, like `-f` means read from file not matter what command you use.
but if @kubernetes/sig-cli-maintainers all agree with this  `-m`, I have no objections.

Comment (liggitt) [reply pkg/kubectl/cmd/drain.go]: > Use short flags sparingly, only for the most frequently used options, prefer lowercase over uppercase for the most common cases, try to stick to well known conventions for UNIX commands and/or Docker, where they exist, and update this list when adding new short flags

This doesn't warrant a short-flag, in my opinion

**Comment by yujuhong** [comment]
I'm not sure about adding a field in the status that can be set by multiple components and users for multiple purposes. Some caveats off the top of my head:
 1. If any component/user did not set/unset the message, you'd be left with stale message that conveys the wrong information.
 2. Some conditions/statuses are not mutually exclusive (e.g., readiness and schedulability). Sharing the same message field would be confusing. 

For the one particular use case -- user-specified message for kubectl operations), maybe an annotation is more suitable?

**Comment by andyxning** [comment]
@yujuhong 

> If any component/user did not set/unset the message, you'd be left with stale message that conveys the wrong information.

For this PR if users do not set `--message` command line argument, by default the `Message` filed will be set to empty which should be fine when node schedulability or readiness change. So, for users, this should not be a problem. But for components,  it should be the problem for most cases as i think most kubernetes components use patch method to update node schedulability or node conditions which should not update the message field if not specified explicitly. 

IMHO, whether we use  annotation or a `Message` field in `Status`, we can not guarantee all components will update the value when node schedulability or readiness changes.

File: pkg/kubectl/cmd/drain.go (hunk 1 above)
Comment (andyxning) [reply pkg/kubectl/cmd/drain.go]: @liggitt @shiywang IIUC, you mean that `--message` is ok but `-m` is not, right?

**Comment by yujuhong** [comment]
> IMHO, whether we use annotation or a Message field in Status, we can not guarantee all components will update the value when node schedulability or readiness changes.

I was suggesting that adding an annotation purely for the case where user wants to set a message. No other components will set/unset this. This solves one particular use case. 

> message is only used to record some helpful info for users to diagnose a node problem instead of every time running some commands and aggregate all possible info to make a conclusion about the problem. So, imho, this maybe ok to combine the possible messages for readiness and schedulability. On the other hand, by splitting readiness and schedulability message into different fields or annotations, it will be confusing to display those messages.

I'm not sure whether having multiple fields would be confusing -- they would each explain the reason behind a node condition. On the other hand, a field that may be set by multiple components and users for multiple purposes seem confusing and less meaningful. As a user, if I want to know why a node is not schedulable, this field may or may not have information relevant to me, or it may tell me why the node is not ready instead. Could you summarize how this `message` field will/should be used in various cases? I think that will help me understand how exactly we plan to use this, and how components are expected to use this field. I'd rather we figure out the use pattern of this field, instead of adding it first, then trying to use it 10 different ways organically.

**Comment by thockin** [comment]
This is going to requires some API consideration.  I agree with @yujuhong 

@kubernetes/api-reviewers

**Comment by andyxning** [comment]
@yujuhong Thanks for the advice. Will summary use cases later.
